### ISSUE TITLE
DER: Dynamically Expandable Representation for Class Incremental Learning

### ISSUE THREAD

**Issue by chaerinkong** [reyllama/MIPAL-CL#6]
## TL; DR

- Expand the model as new task comes, while pruning channel dimension with learnable masks.

## Motivation

- Better stability-plasticity trade-off under memory constraints.

### Expandable Representations

- Simply concat repr. from previous / new feature extractor.
![image](https://user-images.githubusercontent.com/46308407/158138257-d49c15da-26ed-40e5-ad34-46d091cca8a4.png)

- New feature extractor is initialized with previous weights for fast adaptation and forward transfer.
- Train auxiliary classifier that distinguishes btw **current labels and old labels** to encourage new feature extractor to learn novel features (*i.e.,* orthogonal to previous ones)

### Channel-wise Pruning

- Learns channel-wise sigmoid mask with sparsity loss.
![image](https://user-images.githubusercontent.com/46308407/158138602-2798973f-c224-4334-bcee-fdb59e7e81d8.png)
![image](https://user-images.githubusercontent.com/46308407/158138634-57607481-fba8-4f1f-a996-bb83929643bb.png)

- Sparsity loss tries to compress features by making mask components zero, while minimally affecting the performance.

### Final Objective

![image](https://user-images.githubusercontent.com/46308407/158138761-b6d7438f-330c-4cff-9bd5-3e797e4456ac.png)

## Evaluation

![image](https://user-images.githubusercontent.com/46308407/158138801-c74d5644-a81e-4438-8ca7-27e63306443b.png)
![image](https://user-images.githubusercontent.com/46308407/158138818-db911872-ec48-4714-a9d4-62e07361efb1.png)
![image](https://user-images.githubusercontent.com/46308407/158138847-8dea97c2-b721-474a-bdaf-6ed57ec315cf.png)
![image](https://user-images.githubusercontent.com/46308407/158138854-56599e77-4ad4-4b51-adfb-ebf7d07600dd.png)
![image](https://user-images.githubusercontent.com/46308407/158138869-8f196e9d-ad4e-4903-b8ac-072a3560d2e7.png)
